Rpg sidebar tree (see #259)

kugel- opened the pull request:
Continuation of geany#259 since I can't push to @scriptum branch (I suggest you allow that).

I added some cleanup commits on top of @scriptum work and rebased onto 1.33.

@elextr  @b4n  Please give feedback (considering the history in geany#259). I think I'll merge this in maybe 10 days or so.

kugel- commented:
@codebrainz too

elextr commented on doc/geany.txt:
```diff
@@ -530,6 +530,33 @@ order. It is not alphabetical as shown in the documents list
+    Show opened documents as tree similar to directory tree, but it shows only
+    directories that contains opened files. Directories that doesn't contain
+    files are shown as paths (similar to *Show Paths* mode). Home directory
+    always shortened to ``~``. If there is an active project, it makes a tree
```
is always

elextr commented:
@kugel thanks for making a testable version of #259, reviewed but unfortunately I can't test ATM, maybe someone else will test in the meantime.

kugel- commented on doc/geany.txt:
```diff
@@ -530,6 +530,33 @@ order. It is not alphabetical as shown in the documents list
+    files are shown as paths (similar to *Show Paths* mode). Home directory
+    always shortened to ``~``. If there is an active project, it makes a tree
+    of project documents inside project name.
+
```
The node in the tree view is labeled after the project. While normally folders are combined if no files in parent folders are open, this is not done for projects.

e.g. if I open document.c it shows normally as
~/dev/geany.git/src
-- document.c

If I make a project it shows as
geany
-- src
---- document.c

Documents outside the project base dir are not affected (i.e. the nodes are labeled after the paths)

elextr replied:
Oh, ok, so if I read that right only the project base dir and below shows as a tree?

Maybe the selection should be called "Show Project Tree"

Just a final check, commonly my directory layouts look like:

```
/data -- the raid array mount point
    foo -- the project name
        foo -- the git tree and the project base dir in foo.geany
             src
                 foo.cpp
                 lots more source
        foo.geany
        dev_build -- out of tree build dir
              foo -- the executable
              test_output -- unit test output file
              lots more stuff the build makes
```
So if I had project foo open and foo.cpp and test_output open I would get

```
/data/foo/dev_build
      test_output
foo
     src
         foo.cpp
```

Kugel commented:
hey sorry to dissapoint, I dont know how my email got linked to this
project but the Kugel you are looking for doesnt use this email (ive been
getting random updates from github every once  in a while for years), sorry
=( but hey i hope you guys work out those bugs and keep plugging away at it!

On Mon, Mar 26, 2018 at 5:47 PM, elextr <notifications@github.com> wrote:

> *@elextr* commented on this pull request.
> ------------------------------
>
> In doc/geany.txt
> <https://github.com/geany/geany/pull/1813#discussion_r177245479>:
>
> > +
> +    .. image:: ./images/sidebar_documents_only.png
> +
> +Show Paths
> +    Show opened documents as two-level tree in which first level is a full
> +    directory path to document and second level is a file name.
> +
> +    .. image:: ./images/sidebar_show_paths.png
> +
> +Show Tree
> +    Show opened documents as tree similar to directory tree, but it shows only
> +    directories that contains opened files. Directories that doesn't contain
> +    files are shown as paths (similar to *Show Paths* mode). Home directory
> +    always shortened to ``~``. If there is an active project, it makes a tree
> +    of project documents inside project name.
> +
>
> Oh, ok, so if I read that right only the project base dir and below shows
> as a tree?
>
> Maybe the selection should be called "Show Project Tree"
>
> Just a final check, commonly my directory layouts look like:
>
> /data -- the raid array mount point
>     foo -- the project name
>         foo -- the git tree and the project base dir in foo.geany
>              src
>                  foo.cpp
>                  lots more source
>         foo.geany
>         dev_build -- out of tree build dir
>               foo -- the executable
>               test_output -- unit test output file
>               lots more stuff the build makes
>
> So if I had project foo open and foo.cpp and test_output open I would get
>
> /data/foo/dev_build
>       test_output
> foo
>      src
>          foo.cpp
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/geany/geany/pull/1813#discussion_r177245479>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFh94vi_hC13eNhoFXodrBn1SOdsmxTLks5tiWIKgaJpZM4SyLKn>
> .
>

elextr commented:
@kugel sorry, I meant @kugel- but your user comes up first on the autocomplete, and hey us humans make mistakes.

Kugel commented:
hehe its fine, i get these every once in a while, just wanted to make sure
the real kugel gets your message, eve if he stole my nickname =D

On Tue, Mar 27, 2018 at 2:19 AM, elextr <notifications@github.com> wrote:

> @Kugel <https://github.com/Kugel> sorry, I meant @kugel-
> <https://github.com/kugel-> but your user comes up first on the
> autocomplete, and hey us humans make mistakes.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/geany/geany/pull/1813#issuecomment-376410113>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFh94ipxXxoJMJTTXUVW5B-aPy40D7GDks5tidnggaJpZM4SyLKn>
> .
>

kugel- commented:
>Oh, ok, so if I read that right only the project base dir and below
>shows as a tree?

No, all open files are tree'ified. Its just that once a project is open it is ensured that its base path gets its own tree node (labeled after the project)


>
>Maybe the selection should be called "Show Project Tree"


No, see above.

>
>Just a final check, commonly my directory layouts look like:
>
>```
>/data -- the raid array mount point
>    foo -- the project name
>        foo -- the git tree and the project base dir in foo.geany
>             src
>                 foo.cpp
>                 lots more source
>        foo.geany
>        dev_build -- out of tree build dir
>              foo -- the executable
>              test_output -- unit test output file
>              lots more stuff the build makes
>```
>So if I had project foo open and foo.cpp and test_output open I would
>get
>
>```
>/data/foo/dev_build
>      test_output
>foo
>     src
>         foo.cpp
>```


Yes. But if you also had a /data/bar folder then it'd look like 

```
/data
     foo/dev_build
         test_output
     bar
         file.txt
foo
     src
         foo.cpp
```

elextr commented:
> all open files are tree'ified. Its just that once a project is open it is ensured that its base path gets its own tree node (labeled after the project)

Oh, ok, thats sort of unexpected, so it needs to be explained more clearly (if only so we can close issues and point people to the manual that says thats how its meant to work :)

kugel- commented:
Please suggest a better description, I'm not so good at this.

elextr commented:
@kugel- yep, will propose something as soon as I'm happy I understand :)

But on looking at the above example again.

> No, all open files are tree'ified.

But in your additional example you have shown `foo/dev_build` as not being expanded the way `foo/src` is?

kugel- commented:
> But in your additional example you have shown foo/dev_build as not being expanded the way foo/src is?

foo/dev_build is one tree node because no file in foo itself is open. foo->src is two tree nodes because foo represents the project (therefore always gets its own top-level node) and src is a subdirectory (with an open file).

Might be easier if you just try the patch and see yourself, Github/mail kind of sucks for showing examples.

elextr commented:
> Might be easier if you just try the patch and see yourself, Github/mail kind of sucks for showing examples.

Two things:

1. I said above I can't try it ATM

2. guessing how its meant to work from experiments is a poor way of getting an explanation for users

kugel- commented:
@elextr do you need more information on how it's supposed to work?

elextr commented:
Yes, above I said:

> But on looking at the above example again.

> > No, all open files are tree'ified.

> But in your additional example you have shown foo/dev_build as not being expanded the way foo/src is?

I had previously asked if only the project tree was expanded and you said no as quoted above, so to be clear the question is, if "all open files are treeified" then why isn't `foo/dev_build` expanded but `foo/src` is expanded?

kugel- commented:
This proves that I'm not good at documenting such things for end users :-)

There is additional confusion since your example has two foo folders and the project is named foo as well. I'll rename to illustrate:

```
/data -- the raid array mount point
    foo
        bar -- the git tree and the project base dir in baz.geany
             src
                 xx.cpp
                 lots more source
        baz.geany -- project name renamed from foo to baz
        dev_build -- out of tree build dir
              foo -- the executable
              test_output -- unit test output file
              lots more stuff the build makes
```

So accordingly, in your next example foo/src becomes bar/src.

If xx.cpp and test_output are open, then it's displayed as follows:

```
baz
  src
    xx.cpp
/data/foo/dev_build
  test_output
```

This is because of the following rules:
- The directory of a filename is folded as much as possible
- if two or more documents have common parents, then a tree is formed at that parent (repeat for any common parent encountered)
- If a project is open, then the project's base path will get a separate tree node at the root (labeled after the project). This happens regardless of whether parts of the project base path has common parents with open documents outside of the project. Folding of subdirectories (below the project base path) is performed if possible.

Compare to the situation when there is no project open:

elextr commented:
> This proves that I'm not good at documenting such things for end users :-)

Thats ok, thats why I'm asking all these annoying questions :-) 

I think the confusion is that I understood that the difference between this option and the existing option was that all directories on the path to open files expanded, but in fact they only expand if more than one thing inside them is open.  

To confirm by example, if my project `src` directory had three sub-directories `front_end`, `middle_end` and `back_end` then if only a file was open in `front_end` the result for the project tree would be:

```
baz
    src/front_end
        xxx.cpp
```

but if I had a file open in each of `front_end` and `back_end` the result would be:

```
baz
    src
        back_end
            yyy.cpp
        front_end
            xxx.cpp
```
 with `back_end` and `front_end` sorted and `middle_end` not showing.

kugel- commented:
Correct

elextr commented:
@kugel- I still can't do anything with github, so I have posted my suggested text below.

```
Document list views
^^^^^^^^^^^^^^^^^^^

There are three different ways to display documents on the sidebar if *Show
documents list* is active. To switch between views press the right mouse button
on documents list and select one of those radio buttons:

Documents Only
    Show only file names of open documents in sorted order.

    .. image:: ./images/sidebar_documents_only.png

Show Paths
    Show open documents as a two-level tree in which first level is the paths
    of directories containing open files and the second level is the file names of 
    the documents open in that path.  All documents with the same path are grouped 
    together under the same first level item.  Paths are in sorted order and 
    documents are sorted within each group.

    .. image:: ./images/sidebar_show_paths.png

Show Tree
    Show paths as above, but as a multiple level partial tree.  The tree is only 
    expanded at positions where two or more directory paths to open documents 
    share the same prefix.  The common prefix is shown as a parent level, and 
    the remainer of those paths are shown as child levels.  This applies 
    recursively down the paths making a tree to the file names of open documents, 
    which are grouped in sorted order as an additional level below the last path 
    segment.
    
    For convenience two common file locations are handled specifically, open 
    files below the users home directory, and open files below an open project 
    base path.  Each of these is moved to its own top level tree instead of 
    being in place in the normal tree.  The top level of these trees are each 
    labelled differently.  For the home directory tree the path of the home 
    directory is shown as ``~``, and for the project tree the path to the project
    base path is shown simply as the project name.

    .. image:: ./images/sidebar_show_tree.png

In all cases paths and file names that do not fit in the width available are ellipsised.
```

kugel- commented:
Works for me, thanks.

kugel- commented:
I get the following warning when building the manual:

```
[WARNING] image.py:463 image /home/kugel/dev/geany.git/doc/./images/sidebar_documents_only.png is too tall for the frame, rescaling
[WARNING] image.py:463 image /home/kugel/dev/geany.git/doc/./images/sidebar_documents_only.png is too tall for the frame, rescaling
```

In the generated pdf sidebar_documents_only.png appears very small. The HTML manual is fine though. Does anyone know what to do about it?

kugel- commented:
@elextr Actually, the home directory is merely abbreviated (using ~), it doesn't get an individual root node unlike the project base path. See the doc/sidebar_show_tree.png

elextr commented:
@kugel, erm, ok, what happens if there are two directories open along the path to `~` eg if HOME is `/file_server/home/this_user` and I have the files `/file_server/home/this_user/my_file.cpp`, `/file_server/home/this_user/a_dir/also_mine.cpp` and `/file_servert/home/another_user/their_file.cpp` open?

```
~
    a_dir
        also_mine.cpp
    my_file.cpp
/file_server/home/another_user
    their_file.cpp
```
or
```
/file_server/home
    ~
        a_dir
            also_mine.cpp
        my_file.cpp
    another_user
        their_file.cpp
```
or no `~`
```
/file_server/home
    this_user
        a_dir
            also_mine.cpp
        my_file.cpp
    another_user
        their_file.cpp
```

elextr commented:
changed which @kugel-  :)

kugel- commented:
I think the second. However, upon trying I found there is a crash (but only when there are open files of another user AND a file located in ~ itself). I'll dig into it.

On a side note, having a file of another user open seems unlikely, as Geany is a poor program for pair programming :-)

kugel- commented:
I amended the commit message of 7552d5f  due to typos. There are no code changes, except the three new commits on top of @elextr change.

@elextr Upon looking at the code handling the home directory, I think the intention was to match your first case, and it was a bit buggy. Since I think this is the most useful behavior, so I fixed the code to match that behavior.

kugel- commented:
@elextr is your approval still valid even after my changes? I'd like to merge this finally.

elextr commented:
@kugel- I'm fine for you to merge it.

kugel- commented:
@elextr I'm sorry. I found that I have forgotten to publish one more commit I made. Please review this one, though I hope it's welcome one since it simplifies things.

kugel- commented:
@elextr ping

kugel- commented:
@scriptum did you give it a try?

kugel- commented:
@elextr @b4n I really want this in the release so please give it a last look / test or an OK for me to merge it. String freeze is soon. Please!

b4n commented:
@kugel- I'm in the process of reviewing the code, but I might not have a lot of time before Friday.

b4n commented on src/sidebar.c:
```diff
@@ -418,45 +429,319 @@ static gchar *get_doc_folder(const gchar *path)
+			has_parent = gtk_tree_model_iter_parent(model, &iter, &data.best_iter);
+			tree_add_new_dir(parent, has_parent ? &iter : NULL, path);
+			tree_copy_recursive(&data.best_iter, parent);
+			gtk_tree_store_remove(store_openfiles, &data.best_iter);
```
Looking at how it's used, it could be nice to make a `tree_reparent()` or something, which would do this plus removing the old parent.  That the implementation requires copying is sad, but would be encapsulated and wouldn't get people like me to look suspiciously at the calling code before going further :)

kugel- replied:
Looking at this, your suggestion sounds good and would also make the intent of the function more clear.

The copy is currently mainly done because the shortname of a directory node may change (if a new parent is added in-between then child nodes could change from a/b/c/d to a->b/c/d, for example (after opening a/x.c))

kugel- replied:
Oh, and because gtk_tree_store_move_after/_before don't work on sorted tree stores.

b4n commented:
Bug: when I open Geany after this change, so with the default setting: if I right-click on the documents sidebar, every element gets folded. Yeah, even before I click on any action.

More generally fold state seems pretty random when switching between modes.

elextr commented:
I also get all the tree folding up when right clicking.

To be clear, this is unacceptable, I can't close a file without losing my fold state.  

This should be merged as is.

This PR has a history of more bugs being found each time its "ready" probably because @kugel- has only had the chance to work on it intermittently, so although I like the intent I, don't think the implementation is suitable for last minute merging.  

It should be held over and merged after 1.34 (and after its fixed) so any further bugs can be worked out before a release.

kugel- commented:
For some reason I wasn't notified about your updates. I was still thinking review is ongoing. Given the bugs that appeared I agree to post-pone. Thanks for finally testing this

scriptum commented:
> @scriptum did you give it a try?

I tested forked Geany 4 or 3 years ago with this patch. But I didn't tested it on latest Geany versions.

I remember a fold state issue but it is probably common for all GTK applications - GTK doesn't remember children fold state unlike Qt. This feature was planned in GTK 4 or later.

b4n commented:
@scriptum Yes, GTK doesn't maintain tree view fold state below folded elements for you.  However, that doesn't mean elements get folded seemingly randomly :)
The issue I mentioned is not that folding an element and unfolding it doesn't restore children's fold state, but that for some reason when right-clicking (which opens a contextual menu) rows get folded.  Also, keeping fold state when changing mode is something the app will likely have to do, because it changes the hierarchy, so there's no real magic that can be done, it's up to the app to know which new row corresponds to which old one, and thus map fold states if wanted.

elextr commented:
> Also, keeping fold state when changing mode is something the app will likely have to do

Especially so in the cases where a buffer close changes the visible hierarchy in one of the "smart" modes, thats really something that only Geany can do, not GTK.

kugel- commented on src/sidebar.c:
```diff
@@ -418,45 +429,319 @@ static gchar *get_doc_folder(const gchar *path)
+		}
+	}
+	if (data.needle != path)
+		g_free(data.needle);
```
No, you are right, might be a leftover from my modifications over the original version.

kugel- commented on src/sidebar.c:
```diff
@@ -418,45 +429,319 @@ static gchar *get_doc_folder(const gchar *path)
+}
+
+
+static void tree_add_new_dir(GtkTreeIter *child, GtkTreeIter *parent, gchar *file)
```
will fix

kugel- commented on src/sidebar.c:
```diff
@@ -418,45 +429,319 @@ static gchar *get_doc_folder(const gchar *path)
+	GtkTreeModel *model = GTK_TREE_MODEL(store_openfiles);
+
+	path = gtk_tree_model_get_path(model, parent_old);
+	gtk_tree_path_free(path);
```
good catch, seems pretty obvious

kugel- commented:
Trying to finish this up...I could locate and fix the unfold bug and I'm currently work through the other review comments.

kugel- commented on src/sidebar.c:
```diff
@@ -418,45 +429,319 @@ static gchar *get_doc_folder(const gchar *path)
+	gtk_tree_path_free(path);
+	tree_copy_item(&parent, parent_old, parent_new);
+	i = gtk_tree_model_iter_n_children(model, parent_old) - 1;
+	while (i >= 0 && gtk_tree_model_iter_nth_child(model, &child, parent_old, i))
```
The original author probably copied the loop from `on_openfiles_document_action_apply()`. I changed it your suggestion which is a more natual fit.

kugel- commented on src/sidebar.c:
```diff
@@ -418,45 +429,319 @@ static gchar *get_doc_folder(const gchar *path)
+		{
+			/* Even more complicated logic. Both dirs have same
+			 * parent, so create new parent and reparent them */
+			GtkTreeIter parent_buf;
```
Will rename

kugel- commented on src/sidebar.c:
```diff
@@ -1100,7 +1412,7 @@ void sidebar_init(void)
 
 	group = stash_group_new(PACKAGE);
-	stash_group_add_boolean(group, &documents_show_paths, "documents_show_paths", TRUE);
+	stash_group_add_integer(group, (gint*)&documents_show_paths, "documents_show_paths", SHOW_PATHS_LIST);
```
Do you suggest to check every single usage of the variable or is there some kind of point in time after which we can assume the value is sanitized?

kugel- replied:
I added sanitizing in the "load-settings" signal handler.

elextr commented:
@kugel- good to see you get some time, ping when you think its testable.

kugel- commented:
@elextr ping :-) @b4n hopefully I adressed the review comments and didn't forget one. Please test and review.

PS: I merged 1.35 to be up-to-date. I plan to rebase and consolidate a bit before merging (but no complete squash).

elextr commented:
@kugel- I no longer use geany all day every day, so it might be a delay before I can test it, hopefully not too long.

kugel- commented:
@b4n @elextr please test and/or review, thanks :-)

kugel- commented:
@b4n @elextr @codebrainz please test and/or review, thanks :-)

kugel- commented:
@b4n your review is still pending. Please have a look again. I would like to merge this early in the cycle.

ntrel commented on src/sidebar.c:
```diff
@@ -418,45 +429,319 @@ static gchar *get_doc_folder(const gchar *path)
+		TREE_CASE_CHLID_OF,
+		TREE_CASE_PARENT_OF,
+		TREE_CASE_HAVE_SAME_PARENT
+	} best_case;
```
Also, typo CHILD

kugel- commented:
Thanks for the heads up and review!

> To be honest if this works well I don't know why we need the other 2 options besides Show Tree.
Once I used "show tree" I never went back to the other methods. It's really great IMO, especially since it allows to easily close all files within a directory with a simple middle mouse click.

I could imagine that someone might be concerned that the tree gets too many levels, making leafs harder to see (might need to scroll sideways) but I haven't had this problem in multiple years.

kugel- commented:
I did a lot of cleanup and added a logic to remember the fold state of children rows when one of their parents are folded. I would love to merge this soon as I've been using this feature for years now.

I squashed all the commits. The original work of @scriptum is barely recognizable by now so I removed him from the commit history.

elextr commented:
@kugel- sorry, I totally lost track of this PR.  Tested, some observations below, not sure if they are deliberate?  But they were surprising, appeared inconsistent, and in at least one case annoying.

Initial layout:

```
V /data/nexciidoc/attic
    old_spec.edoc
V ~/tmpgeany/geany
    V data
        geany-3.0.css
    V src
        document.c
        > tagmanager
```

The `V` means unrolled, and `>` rolled up, there is a file open in tagmanager but I folded it up.

Note the path starting with `~` sorts after `/data` which is correct because `~` is second last ASCII character

Opened `~/atest.edoc` got: 

```
V ~
    atest.edoc
    V /tmpgeany/geany
        V data
            geany-3.0.css
    V src
        document.c
        > tagmanager
V /data/nexciidoc/attic
    old_spec.edoc
```

Note the resorting of `/data` when `~` became a top level, but the previous top level path still started with `~` and sorted after `/data`.

When I closed `~/atest.edoc`:

```
V ~
    V /tmpgeany/geany
        V data
            geany-3.0.css
        V src
            document.c
            > tagmanager
V /data/nexciidoc/attic
    old_spec.edoc
```

 The `~` stayed separated even though it now had only one entry open.

When I closed `geany-3.0.css`:

I expected:

```
V ~
    V /tmpgeany/geany/src
        document.c
        > tagmanager
V /data/nexciidoc/attic
    old_spec.edoc
```

but got:

```
V ~
    > /tmpgeany/geany/src
V /data/nexciidoc/attic
    old_spec.edoc
```

Note `/tmpgeany/geany/src` folded uncommanded, that would get annoying if it kept happening.

Also `~` still doesn't recombine, but when I closed and reopened Geany it combined and sorted after `/data`.

kugel- commented:
Thanks, I'll check. It's quite hard to get GTK to do what you want. Remembering fold state is not supported by GTK directly so I have to code around it, basically using a hammer so long until it fits.

Do you have comments aside the fold state memory?

elextr commented:
> Do you have comments aside the fold state memory?

The resorting of ~ when it became a standalone level

kugel- commented:
Sorting weirdness seems to be caused by `g_utf8_collate_key_for_filename()` simply ignoring `~`. However, I an acceptable workaround (tread ~ as . for sorting). Now I'm looking why nodes below `~` aren't merged back properly.

elextr commented:
> Sorting weirdness seems to be caused by g_utf8_collate_key_for_filename() simply ignoring ~.

Makes sense, it probably is intended to work on absolute filenames, or at least filenames with the same root (ie directory).

> Now I'm looking why nodes below ~ aren't merged back properly.

Getting ordering and merging stable will probably make saved fold state work a lot easier.

kugel- commented:
@elextr I think I got this right now. Please check.

eht16 commented on src/utils.c:
```diff
@@ -468,7 +468,7 @@ gdouble utils_scale_round(gdouble val, gdouble factor)
 
 /* like g_utf8_strdown() but if @str is not valid UTF8, convert it from locale first.
```
Since `utf8_strdown()` is now public, it should be renamed to `utils_utf8_strdown()` or maybe better `utils_str_utf8_strdown()`.

kugel- replied:
OK

eht16 commented on src/ui_utils.c:
```diff
@@ -68,7 +68,7 @@
 	"scope: %S")
 
-GeanyInterfacePrefs	interface_prefs;
+GEANY_EXPORT_SYMBOL GeanyInterfacePrefs interface_prefs;
```
Why is this necessary?

kugel- replied:
The unit test tests each of the modes, thus needs to modify `interface_prefs.openfiles_path_mode`

eht16 replied:
Ah, I see.

eht16 commented on src/sidebar.c:
```diff
@@ -253,23 +243,161 @@ void sidebar_update_tag_list(GeanyDocument *doc, gboolean update)
 	return cmp;
 }
 
+GEANY_EXPORT_SYMBOL
```
Why is this necessary?
AFAIU `sidebar_create_store_openfiles()` could be even `static`.

kugel- replied:
Again, for the unit test

eht16 commented on src/sidebar.c:
```diff
@@ -409,99 +525,454 @@ static gchar *get_doc_folder(const gchar *path)
 
 /* Also sets doc->priv->iter.
  * This is called recursively in sidebar_openfiles_update_all(). */
+GEANY_EXPORT_SYMBOL
```
Why is `GEANY_EXPORT_SYMBOL` necessary here?

kugel- replied:
Again, for the unit test

eht16 reviewed (changes requested):
Looks good. Had only a few remarks.

kugel- commented on src/sidebar.c:
```diff
@@ -1105,8 +1591,6 @@ void sidebar_init(void)
 	g_signal_connect_after(main_widgets.sidebar_notebook, "switch-page",
 		G_CALLBACK(on_sidebar_switch_page), NULL);
-
-	sidebar_tabs_show_hide(GTK_NOTEBOOK(main_widgets.sidebar_notebook), NULL, 0, NULL);
```
yea

kugel- commented on m4/geany-gtk.m4:
```diff
@@ -0,0 +1,33 @@
+dnl GEANY_CHECK_GTK
+dnl Checks whether the GTK stack is available and new enough. Sets GTK_CFLAGS and GTK_LIBS.
+AC_DEFUN([GEANY_CHECK_GTK],
```
I can do a separate PR if you want. The new `GEANY_CHECK_GTK_FUNCS` is only required for the unit test though

eht16 replied:
Maybe it's also OK then here.

kugel- commented on src/keyfile.c:
```diff
@@ -187,7 +187,7 @@ static void init_pref_groups(void)
 		"tab_close_switch_to_mru", FALSE, "check_tab_close_switch_to_mru");
 	stash_group_add_integer(group, &interface_prefs.tab_pos_sidebar, "tab_pos_sidebar", GTK_POS_TOP);
-	stash_group_add_integer(group, &interface_prefs.documents_show_paths, "documents_show_paths", SHOW_PATHS_LIST);
+	stash_group_add_integer(group, &interface_prefs.openfiles_path_mode, "openfiles_path_mode", -1);
```
I agree. Should I do this as part of this PR?

eht16 replied:
I think so, yes. It's just an item in the `NEWS` file.

kugel- commented on src/libmain.c:
```diff
@@ -1018,20 +1018,12 @@ static const gchar *get_locale(void)
 
 GEANY_EXPORT_SYMBOL
-gint main_lib(gint argc, gchar **argv)
+void main_init_headless(void)
```
This and other changes are necessary for the unit test that I added (2474712b08cb15127a5f61e5caf5e91d6b54582c)

eht16 replied:
Ah, I see.

eht16 commented:
LGTM now! Thanks.

kugel- commented:
Great, thanks! I'll merge in a few days then. @elextr do you plan more tests and/or review or are you fine with this now as well?

elextr commented:
@kugel- Intending to test this weekend.

elextr commented:
@kugel- its better, ~ handling seems more correct, successfully recombines ~ and child when reduced from two children to one.

But folding saving doesn't seem to be right yet.

Initial state

```
v ~
  v old_geany/geany/src
    document.c
  > rpggeany/geany/src  // note this is folded
v /data/misc
  parser.cpp
```

close and re-open Geany gives

```
> ~                // now whole of ~ is folded
v /data/misc
  parser.cpp
```

set back to same initial state and close and re-open, Geany opens with everything unfolded and from then on always seemed to re-open with everything unfolded, no matter what I did.

kugel- commented:
Thanks for testing. I'll check.

Just leaving the note here that folding state is not stored to disk anywhere so re-opening Geany begins anew.

elextr commented:
> Just leaving the note here that folding state is not stored to disk anywhere so re-opening Geany begins anew.

Ok, so opening all unfolded is fine then, so the only issue is the ~ opening folded one time.  Since it didn't repeat the few times I tried maybe its an uninitialised variable that is mostly the correct value, but occasionally bad.

kugel- commented:
I made a minor change that should fix startup (the expectation is that everything is unfolded on startup). Please check again. I couldn't reproduce your second example even without the last correction.

elextr commented:
Ok, now seems consistent, can't cause any bad behaviour, always opens unfolded, ~ coalescing seems to work. :-)

kugel- commented:
I could reproduce the your second example (all under `~` starts up folded). Depends on the order on which documents are opened. I could fix that with another minor tweak. 

So if nothing comes up last minute I will merge by the weekend or so.

elextr commented:
> Depends on the order on which documents are opened. I could fix that with another minor tweak.

Would have been useful to detail the reproducer for testers.

Anyway I found an order that reliably reproduced the problem and your latest push seems to fix it.

> So if nothing comes up last minute I will merge by the weekend or so.

As I have said before, its important to allow a whole weekend before merging, I just happened to have time today, but many people only have weekends to try anything.

kugel- commented:
Merging after CI is done. The last push was just a squash of the fixup (as well as splitting the unit test commit into two). No diff to the previous state.
